### DISCUSSION TITLE
Same app, multiple branches, multiple releases, multiple k8s clusters - best practice ?

### DISCUSSION THREAD

**Discussion by haimari** [fluxcd/flux2#1097] [Q&A]
Hello,

I'll try to describe the situation with as much details as I can.

**Today** I have 3 K8s (On Prem RKE) clusters in different DataCenters. 
In Our Private GitLab(On prem) we have multiple microservices projects.
Each project currently builds only from the "Master" branch. 
In the `release` stage of the pipeline, a `tag` is created as well as a docker image tag with the release version `:v.0.0.2` in the Private GitLab Registry.

On `release` the pipeline runs a stage which goes to a project named 'helm' which holds all helm charts of all projects, and updates the `values.yaml` file of the specific project which was released with the new `image tag`

ArgoCD then picks up the changes in the `helm` repository (apps created with Terraform) and updates the k8s app accordingly.

**What is wrong with the above flow ?**
1. In ArgoCD we have to create the same app 3 times if we want to create it for each cluster. (MultiCluster is not supported)
2. This prevents us from using the `{{ .Release.Name }}` variable generally. (for example for `ingress host path`) forcing us to have too many duplications of configuration files.
3. no GitFlow support. We only support `release` from master branch. other branches cannot be deployed as part of the same app... as they will overwrite the production release... or if we will deploy them as different service, they will not have the same endpoint and will not receive the same traffic, or provide a way to control the amount of traffic.

**What do we want to achieve ?**

1. Single app with **the same name** deployed to **multiple clusters** by the **same Helm chart** with **different values**.
2. Avoid any **duplications** and inherent helm configurations
3. GitFlow support - ability to deploy / canary deploy multiple tags conventions and not replace the entire production app version if for example the `dev` branch is build with new code.

I've been looking at the documentation of v2 and would like to know which if any of the requirements above can be achieved using flux.
Also if there are any Best practices for such GitFlow with multiple clusters.

**Comment by stefanprodan** [comment]
@haimari please take a look at: https://github.com/fluxcd/flux2-kustomize-helm-example

**Reply by haimari** [reply]
Hello @stefanprodan,
Yes I went over this example, yet it shows an example of `staging` and `production` in different clusters. which is great on it's own.

While what i'm looking for (could be out of scope) is to be able to deploy `:dev-1.0.0` & `:dev-1.0.1` & `v:0.9.1` of the same app (from different branches of the app) in the same cluster as the **same** `service` 
(will get the same traffic from the same endpoint address/ingress) 

For example with `ArgoCD`, it's possible to create multiple `Rollouts` and one ` k8s service` which will point to all these `rollouts rs`.

However I'm looking for a better way to support that without managing multiple `Values.yaml` files and CI shell scripts to accomplish that :

multiple clusters --> same namespace --> same app --> multiple tags --> different values for each of those.

**Reply by haimari** [reply]
@stefanprodan 
To better explain this:

Consider I have only a **single** k8s cluster with `prd` & `stg` namespaces (and flux installed on flux-system)
I also have an app which is build as docker with GitLab CI then a docker image is pushed and tagged from the `stg` & `prd` branches:

`stg` branch image tags would be `group/project:vx.x.x-rc`
`prd` branch image tags would be `group/project:vx.x.x`

I would like to use the same `HelmRelease`, `HelmRepository` & `ImageRepository` with different `ImagePolicy` in order to deploy the same app (using the same `HelmRelease`) with different Helm values (for example the `image tag`) in both namespaces of that cluster.

How would such flux directory structure look ?  while considering best GitOps Practices and minimizing code(helm) duplication.

Thanks again

**Reply by scottrigby** [reply]
@haimari do you still want to discuss this? I believe what Stefan meant is you can use Flux objects like HelmRelease together with Kustomize overlays in similar ways to that demo repo, not that the environment directory structure has to match exactly.

I think you have a good use case here – if you would like we could chat about folder structure together, with the intention of documenting this perhaps in a demo repo for this use case. Feel free to reach out on CNCF Slack #flux channel or maybe see you at KubeCon NA if you're going in person.

**Comment by stevendick-work** [comment]
As far as I can see, it's way too early to be talking about 'best practices' with how to setup Flux for different use-cases. We're still discovering what patterns work and work well.

You can define a Flux HelmRelease once and use a per-environment Kustomization patch to make the modifications to the values.

Caring about duplication is a much lower priority than making sure my Flux model/layout works.

My model is more complex as have a separate repo per environment and an environment can optionally be multi-tenant, with the tenant name as part of the target deployment namespace. I also need to deploy the same chart multiple times in an environment with a different release name and config.

All of this is possible with Flux v2, but it requires experimentation with the repo layout and implementation of Flux and K8s Kustomizations.

**Comment by o6uoq** [comment]
This discussion is a few years old now, but interested to hear what patterns have solidifed over the years with relation to the subject. Personally, I prefer to avoid multiple branches and try and shift all logic away from branches to elsewhere in the pipeline. However, subjective and nuanced per requirement and environment.

**Comment by stevendick-work** [comment]
We went the whole way of using a separate Git repo per K8s cluster. This is easier to manage and allows custom rules per repo for access and control. The downside is lack of off-the-shelf tooling for migration/promotion between Git repos.
